SourceDistributions/Developer-Sandboxes/tsWxGTUI_PyVx/Python-2x/tsWxGTUI_Py2x/ManPages/tsManPagesLibGUI/tsWxScrolledWindow.man

Help on module tsWxScrolledWindow:

NAME
    tsWxScrolledWindow

FILE
    ./tsLibGUI/tsWxPkg/src/tsWxScrolledWindow.py

DESCRIPTION
    tsWxScrolledWindow.py - Class to manage scrolling for its
    client area, transforming the coordinates according to the
    scrollbar positions, and setting the scroll positions, thumb
    sizes and ranges according to the area in view.

CLASSES
    tsWxScrolled.Scrolled(tsWxPanel.Panel)
        ScrolledWindow
    
    class ScrolledWindow(tsWxScrolled.Scrolled)
     |  The wxScrolledWindow class manages scrolling for its client area,
     |  transforming the coordinates according to the scrollbar positions,
     |  and setting the scroll positions, thumb sizes and ranges according
     |  to the area in view.
     |  
     |  Starting from version 2.4 of wxWidgets, there are several ways to use
     |  a wxScrolledWindow. In particular, there are now three ways to set the
     |  size of the scrolling area:
     |  
     |  One way is to set the scrollbars directly using a call to
     |  wxScrolledWindow::SetScrollbars. This is the way it used to be in
     |  any previous version of wxWidgets and it will be kept for backwards
     |  compatibility.
     |  
     |  An additional method of manual control, which requires a little less
     |  computation of your own, is to set the total size of the scrolling
     |  area by calling either wxWindow::SetVirtualSize, or wxWindow::FitInside,
     |  and setting the scrolling increments for it by calling
     |  wxScrolledWindow::SetScrollRate. Scrolling in some orientation is
     |  enabled by setting a non-zero increment for it.
     |  
     |  The most automatic and newest way is to simply let sizers determine
     |  the scrolling area. This is now the default when you set an interior
     |  sizer into a wxScrolledWindow with wxWindow::SetSizer. The scrolling
     |  area will be set to the size requested by the sizer and the scrollbars
     |  will be assigned for each orientation according to the need for them
     |  and the scrolling increment set by wxScrolledWindow::SetScrollRate.
     |  As above, scrolling is only enabled in orientations with a non-zero
     |  increment. You can influence the minimum size of the scrolled area
     |  controlled by a sizer by calling wxWindow::SetVirtualSizeHints.
     |  (calling wxScrolledWindow::SetScrollbars has analogous effects in
     |  wxWidgets 2.4 -- in later versions it may not continue to override
     |  the sizer)
     |  
     |  Note: if Maximum size hints are still supported by SetVirtualSizeHints,
     |  use them at your own dire risk. They may or may not have been removed
     |  for 2.4, but it really only makes sense to set minimum size hints here.
     |  We should probably replace SetVirtualSizeHints with SetMinVirtualSize
     |  or similar and remove it entirely in future.
     |  
     |  As with all windows, an application can draw onto a wxScrolledWindow
     |  using a device context.
     |  
     |  You have the option of handling the OnPaint handler or overriding the
     |  OnDraw function, which is passed a pre-scrolled device context
     |  (prepared by DoPrepareDC).
     |  
     |  If you do not wish to calculate your own scrolling, you must call
     |  DoPrepareDC when not drawing from within OnDraw, to set the device
     |  origin for the device context according to the current scroll position.
     |  
     |  A wxScrolledWindow will normally scroll itself and therefore its
     |  child windows as well. It might however be desired to scroll a
     |  different window than itself: e.g. when designing a spreadsheet,
     |  you will normally only have to scroll the (usually white) cell area,
     |  whereas the (usually grey) label area will scroll very differently.
     |  For this special purpose, you can call SetTargetWindow which means
     |  that pressing the scrollbars will scroll a different window.
     |  
     |  Note that the underlying system knows nothing about scrolling
     |  coordinates, so that all system functions (mouse events, expose events,
     |  refresh calls etc) as well as the position of subwindows are relative
     |  to the "physical" origin of the scrolled window. If the user insert
     |  a child window at position (10,10) and scrolls the window down 100
     |  pixels (moving the child window out of the visible area), the child
     |  window will report a position of (10,-90).
     |  
     |  Derived from
     |  
     |  wxPanel
     |  wxWindow
     |  wxEvtHandler
     |  wxObject
     |  
     |  Method resolution order:
     |      ScrolledWindow
     |      tsWxScrolled.Scrolled
     |      tsWxPanel.Panel
     |      tsWxWindow.Window
     |      tsWxEvtHandler.EvtHandler
     |      tsWxObject.Object
     |      __builtin__.object
     |  
     |  Methods defined here:
     |  
     |  __init__(self, parent, id=-1, pos=(-1, -1), size=(-1, -1), style=3221225472L, name='panel')
     |      Construct and show a generic Window.
     |  
     |  ----------------------------------------------------------------------
     |  Methods inherited from tsWxScrolled.Scrolled:
     |  
     |  AdjustScrollbars(self)
     |      TBD - Under Construction.
     |  
     |  CalcScrollInc(self, event)
     |      TBD - Under Construction.
     |  
     |  CalcScrolledPosition(self, x, y, xx, yy)
     |      Translates the logical coordinates to the device ones.
     |      For example, if a window is scrolled 10 pixels to the bottom,
     |      the device coordinates of the origin are (0, 0) (as always),
     |      but the logical coordinates are (0, 10) and so the call to
     |      CalcScrolledPosition(0, 10, &xx, &yy) will return 0 in yy.
     |  
     |  CalcUnscrolledPosition(self, x, y, xx, yy)
     |      Translates the device coordinates to the logical ones. For
     |      example, if a window is scrolled 10 pixels to the bottom,
     |      the device coordinates of the origin are (0, 0) (as always),
     |      but the logical coordinates are (0, 10) and so the call to
     |      CalcUnscrolledPosition(0, 0, &xx, &yy) will return 10 in yy.
     |  
     |  Create(self, parent, id=1, pos=(-1, -1), size=(-1, -1), style=3221225472L, name='panel', pixels=True)
     |      Creates the window for two-step construction. Derived classes
     |      should call or replace this function.
     |  
     |  DisableKeyboardScrolling(self)
     |      Disable use of keyboard keys for scrolling.
     |      
     |      By default cursor movement keys (including Home, End, Page Up and
     |      Down) are used to scroll the window appropriately. If the derived
     |      class uses these keys for something else, e.g. changing the
     |      currently selected item, this function can be used to disable this
     |      behaviour as it is not only not necessary then but can actually
     |      be actively harmful if another object forwards a keyboard event
     |      corresponding to one of the above keys to us using
     |      ProcessWindowEvent() because the event will always be processed
     |      which can be undesirable.
     |  
     |  DoPrepareDC(self, dc)
     |      Call this function to prepare the device context for drawing a
     |      scrolled image. It sets the device origin according to the current
     |      scroll position.
     |      
     |      DoPrepareDC is called automatically within the default
     |      wxScrolledWindow::OnPaint event handler, so your
     |      wxScrolledWindow::OnDraw override will be passed a pre-scrolled
     |      device context. However, if you wish to draw from outside of
     |      OnDraw (via OnPaint), or you wish to implement OnPaint yourself,
     |      you must call this function yourself.
     |  
     |  EnableScrolling(self, x_scrolling=False, y_scrolling=False)
     |      Enable or disable physical scrolling in the given direction.
     |      Physical scrolling is the physical transfer of bits up or down
     |      the screen when a scroll event occurs. If the application
     |      scrolls by a variable amount (e.g. if there are different font
     |      sizes) then physical scrolling will not work, and you should
     |      switch it off. Note that you will have to reposition child
     |      windows yourself, if physical scrolling is disabled.
     |      
     |      Parameters
     |      
     |      xScrolling
     |      
     |      If true, enables physical scrolling in the x direction.
     |      yScrolling
     |      
     |      If true, enables physical scrolling in the y direction.
     |      Remarks
     |      
     |      Physical scrolling may not be available on all platforms.
     |      Where it is available, it is enabled by default.
     |  
     |  GetScaleX(self)
     |      Return the ScaleX status of tsWxScrolledText.
     |  
     |  GetScaleY(self)
     |      Return the ScaleY status of tsWxScrolledText.
     |  
     |  GetScrollPageSize(self, orient)
     |      Return the ScalePageSize status of tsWxScrolledText for the
     |      designated orientation.
     |  
     |  GetScrollPixelsPerUnit(self)
     |      Get the number of pixels per scroll unit (line) of tsWxScrolledText,
     |      in each direction, as set by wxScrolledWindow::SetScrollbars. A
     |      value of zero indicates no scrolling in that direction.
     |      
     |      Parameters
     |      
     |      xUnit
     |      
     |      Receives the number of pixels per horizontal unit.
     |      yUnit
     |      
     |      Receives the number of pixels per vertical unit.
     |  
     |  GetScrollRect(self)
     |      Return the Scrolled Window Rectangle of tsWxScrolledText.
     |  
     |  GetSizeAvailableForScrollTarget(self)
     |      Function which must be overridden to implement the size available
     |      for the scroll target for the given size of the main window.
     |      
     |      This method must be overridden if SetTargetWindow() is used
     |      (it is never called otherwise). The implementation should decrease
     |      the size to account for the size of the non-scrollable parts of the
     |      main window and return only the size available for the scrollable
     |      window itself. E.g. in the example given in SetTargetWindow()
     |      documentation the function would subtract the height of the
     |      header window from the vertical component of size.
     |  
     |  GetTargetRect(self)
     |      Return Target Rectangle.
     |  
     |  GetTargetSize(self)
     |      Return Target Size.
     |  
     |  GetTargetWindow(self)
     |      Return Target Window.
     |  
     |  GetViewStart(self)
     |      Get the position at which the visible portion of the window starts.
     |      
     |      Parameters
     |      
     |      x - Receives the first visible x position in scroll units.
     |      y - Receives the first visible y position in scroll units.
     |  
     |  GetVirtualSize(self)
     |      Gets the size in device units of the scrollable window area
     |      (as opposed to the client size, which is the area of the
     |      window currently visible).
     |      
     |      Parameters
     |      
     |      x - Receives the length of the scrollable window, in pixels.
     |      y - Receives the height of the scrollable window, in pixels.
     |  
     |  IsAutoScrolling(self)
     |      Return True only if AutoScrolling is currently active.
     |  
     |  IsRetained(self)
     |      Motif only: true if the window has a backing bitmap.
     |  
     |  OnDraw(self, dc)
     |      Called by the default paint event handler to allow the
     |      application to define painting behaviour without having to
     |      worry about calling wxScrolledWindow::DoPrepareDC.
     |      
     |      Instead of overriding this function you may also just process
     |      the paint event in the derived class as usual, but then you
     |      will have to call DoPrepareDC() yourself.
     |  
     |  PrepareDC(self, dc)
     |      This function is for backwards compatibility only and simply
     |      calls DoPrepareDC now. Notice that it is not called by the
     |      default paint event handle (DoPrepareDC() is), so overriding
     |      this method in your derived class is useless.
     |  
     |  Scroll(self, x_pos, y_pos)
     |      Scrolls a window so the view start is at the given point.
     |      
     |      Parameters
     |      
     |      x_pos - The x position to scroll to, in scroll units.
     |      
     |      y_pos - The y position to scroll to, in scroll units.
     |      
     |      Remarks
     |      
     |      The positions are in scroll units, not pixels, so to convert
     |      to pixels you will have to multiply by the number of pixels
     |      per scroll increment. If either parameter is -1, that position
     |      will be ignored (no change in that direction).
     |      
     |      Column 0, Row 0 Position (x_pos, y_pos) = (0, 0) is left-most
     |      column on top-most row.
     |      
     |      Column 1, Row 1 Position (x_pos, y_pos) = (1, 1) is one column
     |      to the right and one row down.
     |  
     |  SetScale(self, xs, ys)
     |      TBD - Under Construction.
     |  
     |  SetScrollPageSize(self, orient, pageSize)
     |      TBD - Under Construction.
     |  
     |  SetScrollRate(self, xstep, ystep)
     |      Set the horizontal and vertical scrolling increment only.
     |      See the pixelsPerUnit parameter in SetScrollbars.
     |  
     |  SetScrollbars(self, pixelsPerUnitX, pixelsPerUnitY, noUnitsX, noUnitsY, xPos=0, yPos=0, noRefresh=False)
     |      Sets up vertical and/or horizontal scrollbars.
     |      
     |      Parameters
     |      
     |      pixelsPerUnitX
     |      Pixels per scroll unit in the horizontal direction.
     |      
     |      pixelsPerUnitY
     |      Pixels per scroll unit in the vertical direction.
     |      
     |      noUnitsX
     |      Number of units in the horizontal direction.
     |      
     |      noUnitsY
     |      Number of units in the vertical direction.
     |      
     |      xPos
     |      Position to initialize the scrollbars in the horizontal
     |      direction, in scroll units.
     |      
     |      yPos
     |      Position to initialize the scrollbars in the vertical
     |      direction, in scroll units.
     |      
     |      noRefresh
     |      Will not refresh window if true.
     |      
     |      Remarks
     |      
     |      The first pair of parameters give the number of pixels per
     |      scroll step, i.e. amount moved when the up or down scroll
     |      arrows are pressed. The second pair gives the length of
     |      scrollbar in scroll steps, which sets the size of the virtual window.
     |      
     |      xPos and yPos optionally specify a position to scroll to immediately.
     |      
     |      For example, the following gives a window horizontal and vertical
     |      scrollbars with 20 pixels per scroll step, and a size of 50 steps
     |      (1000 pixels) in each direction.
     |      
     |      window->SetScrollbars(20, 20, 50, 50);
     |      
     |      wxScrolledWindow manages the page size itself, using the current
     |      client window size as the page size.
     |      
     |      Note that for more sophisticated scrolling applications, for
     |      example where scroll steps may be variable according to the
     |      position in the document, it will be necessary to derive a
     |      new class from wxWindow, overriding OnSize and adjusting the
     |      scrollbars appropriately.
     |  
     |  SetTargetWindow(self, target)
     |      Call this function to tell wxScrolled to perform the actual
     |      scrolling on a different window (and not on itself).
     |      
     |      This method is useful when only a part of the window should be
     |      scrolled. A typical example is a control consisting of a fixed
     |      header and the scrollable contents window: the scrollbars are
     |      attached to the main window itself, hence it, and not the
     |      contents window must be derived from wxScrolled, but only
     |      the contents window scrolls when the scrollbars are used.
     |      To implement such setup, you need to call this method with
     |      the contents window as argument.
     |      
     |      Notice that if this method is used,
     |      GetSizeAvailableForScrollTarget() method must be overridden.
     |  
     |  ShowScrollbars(self, horiz=False, vert=False)
     |      Set the scrollbar visibility.
     |      
     |      By default the scrollbar in the corresponding direction is only
     |      shown if it is needed, i.e. if the virtual size of the scrolled
     |      window in this direction is greater than the current physical
     |      window size. Using this function the scrollbar visibility can be
     |      changed to be:
     |      
     |      wxSHOW_SB_ALWAYS: To always show the scrollbar, even if it is not
     |      needed currently (wxALWAYS_SHOW_SB style can be used during the
     |      window creation to achieve the same effect but it applies in both
     |      directions).
     |      
     |      wxSHOW_SB_NEVER: To never show the scrollbar at all. In this case
     |      the program should presumably provide some other way for the user
     |      to scroll the window.
     |      
     |      wxSHOW_SB_DEFAULT: To restore the default behaviour described above.
     |      
     |      Parameters:
     |      horz    The desired visibility for the horizontal scrollbar.
     |      vert    The desired visibility for the vertical scrollbar.
     |  
     |  tsAppendText(self, text, markup=None)
     |              Appends the text argument to the end of the text in the control.
     |              The insertion point also moves to the end of the control.
     |      
     |              Parameters:
     |      
     |              text - String of ascii text characters. Accepted values include
     |                     empty string, simple string, and compound string separated
     |                     by embedded "
     |      ".
     |      
     |              markup - curses text markup ored with color pair number
     |  
     |  tsGet_AutoScrolling(self)
     |      Return the AutoScrolling status of tsWxScrolledText.
     |  
     |  tsGet_MaxTextCols(self)
     |      Return the MaxTextCols status of tsWxScrolledText.
     |  
     |  tsGet_MaxTextRows(self)
     |      Return the MaxTextRows status of tsWxScrolledText.
     |  
     |  tsGet_Retained(self)
     |      Return the Retained status of tsWxScrolledText.
     |  
     |  tsGet_ScaleX(self)
     |      Return the ScaleX status of tsWxScrolledText.
     |  
     |  tsGet_ScaleY(self)
     |      Return the ScaleY status of tsWxScrolledText.
     |  
     |  tsGet_ScrolledTextLines(self)
     |      Return the ScrolledTextLines status of tsWxScrolledText.
     |  
     |  tsGet_TargetWindow(self)
     |      Return the TargetWindow status of tsWxScrolledText.
     |  
     |  tsGet_ViewStart(self)
     |      Return the ViewStart status of tsWxScrolledText.
     |  
     |  tsGet_xScrollLines(self)
     |      Return the xScrollLines status of tsWxScrolledText.
     |  
     |  tsGet_xScrollLinesPerPage(self)
     |      Return the xScrollLinesPerPage status of tsWxScrolledText.
     |  
     |  tsGet_xScrollPixelsPerLine(self)
     |      Return the xScrollPixelsPerLine status of tsWxScrolledText.
     |  
     |  tsGet_xScrollPosition(self)
     |      Return the xScrollPosition status of tsWxScrolledText.
     |  
     |  tsGet_xScrollingEnabled(self)
     |      Return the xScrollingEnabled status of tsWxScrolledText.
     |  
     |  tsGet_yScrollLines(self)
     |      Return the yScrollLines status of tsWxScrolledText.
     |  
     |  tsGet_yScrollLinesPerPage(self)
     |      Return the yScrollLinesPerPage status of tsWxScrolledText.
     |  
     |  tsGet_yScrollPixelsPerLine(self)
     |      Return the yScrollPixelsPerLine status of tsWxScrolledText.
     |  
     |  tsGet_yScrollPosition(self)
     |      Return the yScrollPosition status of tsWxScrolledText.
     |  
     |  tsGet_yScrollingEnabled(self)
     |      Return the yScrollingEnabled status of tsWxScrolledText.
     |  
     |  tsOnHScrollWin(self, evt)
     |  
     |  tsOnHScrollWinBottom(self, evt)
     |      Set scroll position to page containing right-most column.
     |  
     |  tsOnHScrollWinLineDown(self, evt)
     |      Set scroll position to next column.
     |  
     |  tsOnHScrollWinLineUp(self, evt)
     |      Set scroll position to previous column.
     |  
     |  tsOnHScrollWinPageDown(self, evt)
     |      Set scroll position to next page column.
     |  
     |  tsOnHScrollWinPageUp(self, evt)
     |      Set scroll position to previous page column.
     |  
     |  tsOnHScrollWinThumbRelease(self, evt)
     |  
     |  tsOnHScrollWinThumbTrack(self, evt)
     |  

Documents/NEWS.txt
#-----------------------------------------------------------
#"Time-stamp: <07/08/2015  4:48:41 AM rsg>"
#-----------------------------------------------------------

========================= NEWS.txt =========================

   +----+----+  TeamSTARS "tsWxGTUI_PyVx" Toolkit
   | ts | Wx |      with Python 2x & Python 3x based
   +----+----+         Command Line Interface (CLI)
   | G T U I |      and "Curses"-based "wxPython"-style, 
   +---------+         Graphical-Text User Interface (GUI)
 
   Get that cross-platform, pixel-mode "wxPython" feeling
   on character-mode 8-/16-color (xterm-family) & non-color
   (vt100-family) terminals and terminal emulators.

   You can find this and other plain-text files in the
   Toolkit subdirectory named:

       "./<Toolkit Recipient's Repository>/Documents".

       <Your Working Repository>
       (e.g. "tsWxGTUI_PyVx_Repository") 
         |
         |
         +-- ["Documents"]

=========================== News ===========================

The pre-alpha release of the TeamSTARS "tsWxGTUI_PyVx-0.0.0"
Toolkit is now available. It includes the following two
versions abd their associated variants (Python Distribution
Utilities ("Distutils") installable "Site-Packages" and non-
installable "Developer-Sandboxes" for enhancement and bug
fix development and experimentation without interference
with the already installed Site-Packages):

1. The TeamSTARS "tsWxGTUI_Py2x-0.0.0" Toolkit is released
   for users of the popular, high-level, cross-platform,
   second generation Python 2.x programming language which
   is widely available.

   A wiki posting by the Python Software Foundation announc-
   ed that Python 3.0 was released in 2008. The final 2.x
   version 2.7 release came out in mid-2010, with a state-
   ment of extended support for this end-of-life release.
   The 2.x branch will see no new major releases after that.

   The source code and documentation is available for down-
   load at:

      http://pypi.python.org/pypi.

   It supports development and operation of:

   a. Command Line User Interface (CLI) application
      programs implemented in Python 2.0.0-2.7.9.

   b. Graphical-style User Interface (GUI) application
      programs implemented in Python 2.6.4-2.7.9.

2. The TeamSTARS "tsWxGTUI_Py3x-0.0.0" Toolkit is released
   for users of the popular, high-level, cross-platform,
   third generation Python 3.x programming language which
   is growing in availability.

   A wiki posting by the Python Software Foundation announc-
   ed that Python 3.0 was released in 2008. 3.x is under
   active development and has already seen over five years
   of stable releases, including version 3.3 in 2012 and
   3.4 in 2014. This means that all recent standard library
   improvements, for example, are only available by default
   in Python 3.x.

   Guido van Rossum (the original creator of the Python
   language) decided to clean up Python 2.x properly,
   with less regard for backwards compatibility than is
   the case for new releases in the 2.x range. The most
   drastic improvement is the better Unicode support
   (with all text strings being Unicode by default) as
   well as saner bytes/Unicode separation.

   Besides, several aspects of the core language (such as
   print and exec being statements, integers using floor
   division) have been adjusted to be easier for newcomers
   to learn and to be more consistent with the rest of the
   language, and old cruft has been removed (for example,
   all classes are now new-style, "range()" returns a memory
   efficient iterable, not a list as in 2.x).

   The source code and documentation is available for down-
   load at:

      http://pypi.python.org/pypi.

   It supports development and operation of:

   a. Command Line User Interface (CLI) application
      programs implemented in Python 3.0.0-3.4.2.

   b. Graphical-style User Interface (GUI) application
      programs implemented in Python 3.1.5-3.4.2.

======================= End-Of-File ========================


SourceDistributions/Developer-Sandboxes/tsWxGTUI_PyVx/Python-2x/tsWxGTUI_Py2x/tsDemoArchive/tsTutorialDeveloperSandbox2x/tsLibGUI/tsWxPkg/src/tsWxScrolledWindow.py
#! /usr/bin/env python
# "Time-stamp: <10/25/2013  6:47:31 AM rsg>"
'''
tsWxScrolledWindow.py - Class to manage scrolling for its
client area, transforming the coordinates according to the
scrollbar positions, and setting the scroll positions, thumb
sizes and ranges according to the area in view.
'''
#################################################################
#
# File: tsWxScrolledWindow.py
#
# Purpose:
#
#    Class to manage scrolling for its client area, transforming
#    the coordinates according to the scrollbar positions, and
#    setting the scroll positions, thumb sizes and ranges
#    according to the area in view.
#
# Usage (example):
#
#    # Import
#
#    from tsWxScrolledWindow import ScrolledWindow as wxScrolledWindow
#
# Requirements:
#
#    wxPython 2.8.9.2 (New wxPyDocs)/wxWidgets 2.9.2 (Ref. Man.)
#
# Capabilities:
#
#    wxPython 2.8.9.2 (New wxPyDocs)/wxWidgets 2.9.2 (Ref. Man.)
#
# Limitations:
#
#    wxPython 2.8.9.2 (New wxPyDocs)/wxWidgets 2.9.2 (Ref. Man.)
#
# Notes:
#
#    See Notes section in tsWxScrolled.py
#
# Methods:
#
#    wxPython 2.8.9.2 (New wxPyDocs)/wxWidgets 2.9.2 (Ref. Man.)
#
# Classes:
#
#    wxPython 2.8.9.2 (New wxPyDocs)/wxWidgets 2.9.2 (Ref. Man.)
#
# Modifications:
#
#    2012/03/01 rsg Replaced wxPython-style logic with those of
#                   the base class, tsWxScrolled.
#
# ToDo:
#
#    None.
#
#################################################################

__title__     = 'tsWxScrolledWindow'
__version__   = '1.0.0'
__date__      = '04/01/2013'
__authors__   = 'Richard S. Gordon'
__copyright__ = 'Copyright (c) 2007-2013 ' + \
                '%s.\n\t\tAll rights reserved.' % __authors__
__license__   = 'GNU General Public License, ' + \
                'Version 3, 29 June 2007'
__credits__   = '\n\n  Credits: ' + \
                '\n\n\t  tsLibGUI Import & Application Launch Features: ' + \
                '\n\t  Copyright (c) 2007-2009 Frederick A. Kier.' + \
                '\n\t\t\tAll rights reserved.' + \
                '\n\n\t  Python Curses Module API & ' + \
                'Run Time Library Features:' + \
                '\n\t  Copyright (c) 2001-2013 ' +\
                'Python Software Foundation.' + \
                '\n\t\t\tAll rights reserved.' + \
                '\n\t  PSF License Agreement for Python 2.7.3 & 3.3.0' + \
                '\n\n\t  wxWidgets (formerly wxWindows) & ' + \
                'wxPython API Features:' + \
                '\n\t  Copyright (c) 1992-2008 Julian Smart, ' + \
                'Robert Roebling,' + \
                '\n\t\t\tVadim Zeitlin and other members of the ' + \
                '\n\t\t\twxWidgets team.' + \
                '\n\t\t\tAll rights reserved.' + \
                '\n\t  wxWindows Library License' + \
                '\n\n\t  nCurses API & Run Time Library Features:' + \
                '\n\t  Copyright (c) 1998-2011 ' + \
                'Free Software Foundation, Inc.' + \
                '\n\t\t\tAll rights reserved.' + \
                '\n\t  GNU General Public License, ' + \
                'Version 3, 29 June 2007'

__line1__ = '%s, v%s (build %s)' % (__title__, __version__, __date__)
__line2__ = 'Author(s): %s' % __authors__
__line3__ = '%s' % __copyright__

if len( __credits__) == 0:
    __line4__ = '%s' % __license__
else:
    __line4__ = '%s%s' % (__license__, __credits__)

__header__ = '\n\n%s\n\n  %s\n  %s\n  %s\n' % (__line1__,
                                               __line2__,
                                               __line3__,
                                               __line4__)

mainTitleVersionDate = __line1__

## import tsExceptions as tse

import tsWxGlobals as wx
from tsWxScrolled import Scrolled

#---------------------------------------------------------------------------

##DEBUG = True
DEBUG = wx.Debug_GUI_Launch | \
        wx.Debug_GUI_Progress | \
        wx.Debug_GUI_Termination | \
        wx.Debug_GUI_Exceptions

##VERBOSE = True
VERBOSE = wx.Debug_GUI_Configuration

#---------------------------------------------------------------------------

class ScrolledWindow(Scrolled):
    '''
    The wxScrolledWindow class manages scrolling for its client area,
    transforming the coordinates according to the scrollbar positions,
    and setting the scroll positions, thumb sizes and ranges according
    to the area in view.

    Starting from version 2.4 of wxWidgets, there are several ways to use
    a wxScrolledWindow. In particular, there are now three ways to set the
    size of the scrolling area:

    One way is to set the scrollbars directly using a call to
    wxScrolledWindow::SetScrollbars. This is the way it used to be in
    any previous version of wxWidgets and it will be kept for backwards
    compatibility.

    An additional method of manual control, which requires a little less
    computation of your own, is to set the total size of the scrolling
    area by calling either wxWindow::SetVirtualSize, or wxWindow::FitInside,
    and setting the scrolling increments for it by calling
    wxScrolledWindow::SetScrollRate. Scrolling in some orientation is
    enabled by setting a non-zero increment for it.

    The most automatic and newest way is to simply let sizers determine
    the scrolling area. This is now the default when you set an interior
    sizer into a wxScrolledWindow with wxWindow::SetSizer. The scrolling
    area will be set to the size requested by the sizer and the scrollbars
    will be assigned for each orientation according to the need for them
    and the scrolling increment set by wxScrolledWindow::SetScrollRate.
    As above, scrolling is only enabled in orientations with a non-zero
    increment. You can influence the minimum size of the scrolled area
    controlled by a sizer by calling wxWindow::SetVirtualSizeHints.
    (calling wxScrolledWindow::SetScrollbars has analogous effects in
    wxWidgets 2.4 -- in later versions it may not continue to override
    the sizer)

    Note: if Maximum size hints are still supported by SetVirtualSizeHints,
    use them at your own dire risk. They may or may not have been removed
    for 2.4, but it really only makes sense to set minimum size hints here.
    We should probably replace SetVirtualSizeHints with SetMinVirtualSize
    or similar and remove it entirely in future.

    As with all windows, an application can draw onto a wxScrolledWindow
    using a device context.

    You have the option of handling the OnPaint handler or overriding the
    OnDraw function, which is passed a pre-scrolled device context
    (prepared by DoPrepareDC).

    If you do not wish to calculate your own scrolling, you must call
    DoPrepareDC when not drawing from within OnDraw, to set the device
    origin for the device context according to the current scroll position.

    A wxScrolledWindow will normally scroll itself and therefore its
    child windows as well. It might however be desired to scroll a
    different window than itself: e.g. when designing a spreadsheet,
    you will normally only have to scroll the (usually white) cell area,
    whereas the (usually grey) label area will scroll very differently.
    For this special purpose, you can call SetTargetWindow which means
    that pressing the scrollbars will scroll a different window.

    Note that the underlying system knows nothing about scrolling
    coordinates, so that all system functions (mouse events, expose events,
    refresh calls etc) as well as the position of subwindows are relative
    to the "physical" origin of the scrolled window. If the user insert
    a child window at position (10,10) and scrolls the window down 100
    pixels (moving the child window out of the visible area), the child
    window will report a position of (10,-90).

    Derived from

    wxPanel
    wxWindow
    wxEvtHandler
    wxObject
    '''

    #-----------------------------------------------------------------------

    def __init__(self,
                 parent,
                 id=wx.ID_ANY,
                 pos=wx.DefaultPosition,
                 size=wx.DefaultSize,
                 style=wx.HSCROLL | wx.VSCROLL,
                 name=wx.PanelNameStr):
        '''
        Construct and show a generic Window.
        '''
        theClass = 'ScrolledWindow'

        # Capture initial caller parametsrs before they are changed
        self.caller_parent = parent
        self.caller_id = id
        self.caller_pos = pos
        self.caller_size = size
        self.caller_style = style
        self.caller_name = name

        wx.RegisterFirstCallerClassName(self, theClass)

        Scrolled.__init__(self,
                          parent,
                          id=id,
                          pos=pos,
                          size=size,
                          style=style,
                          name=name)

        self.tsBeginClassRegistration(theClass, id)

        self.ts_Name = name
        self.ts_Parent = parent

        theTopLevelClass = self
        self.SetTopLevelAncestor(theTopLevelClass)

        if parent is None:
            self.ts_GrandParent = None
        else:
            self.ts_GrandParent = parent.Parent

        self.ts_BackgroundColour = wx.ThemeToUse['ForegroundColour'].lower()
        self.ts_ForegroundColour = wx.ThemeToUse['BackgroundColour'].lower()

        myRect, myClientRect = self.tsScrolledLayout(
            parent, pos, size, style, name)
        self.ts_Rect = myRect
        self.ts_ClientRect = myClientRect

        thePosition = self.Position
        theSize = self.Size

        if DEBUG:
            self.logger.debug('               self: %s' % self)
            self.logger.debug('             parent: %s' % parent)
            self.logger.debug('                 id: %s' % self.ts_Id)
            self.logger.debug('         AssignedId: %s' % self.ts_AssignedId)
            self.logger.debug('                pos: %s' % thePosition)
            self.logger.debug('               size: %s' % theSize)
            self.logger.debug('              style: 0x%X' % style)
            self.logger.debug('               name: %s' % name)

        self.tsEndClassRegistration(theClass)

    #-----------------------------------------------------------------------
    # Begin tsWx API Extensions

    #-----------------------------------------------------------------------

    # End tsWx API Extensions
    #-----------------------------------------------------------------------

#---------------------------------------------------------------------------

#---------------------------------------------------------------------------

if __name__ == '__main__':

    print(__header__)

